src/main/java/de/tum/in/www1/artemis/service/connectors/localci/buildagent/BuildJobContainerService.java

```diff
@@ -189,6 +190,51 @@ public void stopContainer(String containerName) {
         // for some other reason.
         // The container's main process is waiting for this file to appear and then stops the main process, thus stopping and removing the container.
         executeDockerCommandWithoutAwaitingResponse(containerId, "touch", LOCALCI_WORKING_DIRECTORY + "/stop_container.txt");
+
+        // If the container is still running, force kill it.
+        containers = dockerClient.listContainersCmd().withShowAll(true).exec();
+        containerOptional = containers.stream().filter(container -> container.getNames()[0].equals("/" + containerName)).findFirst();
+        containerOptional.ifPresent(container -> dockerClient.removeContainerCmd(container.getId()).withForce(true).exec());
+    }
+
+    /**
+     * Stops or kills a container in case a build job has failed or the container is unresponsive.
+     * Adding a file "stop_container.txt" like in {@link #stopContainer(String)} might not work for unresponsive containers, thus we use
+     * {@link DockerClient#stopContainerCmd(String)} and {@link DockerClient#killContainerCmd(String)} to stop or kill the container.
+     *
+     * @param containerId The ID of the container to stop or kill.
+     */
+    public void stopUnresponsiveContainer(String containerId) {
+        try {
+            // Attempt to stop the container. It should stop the container and auto-remove it.
+            // {@link DockerClient#stopContainerCmd(String)} first sends a SIGTERM command to the container to gracefully stop it,
+            // and if it does not stop within the timeout, it sends a SIGKILL command to kill the container.
+            dockerClient.stopContainerCmd(containerId).withTimeout(5).exec();
+        }
+        catch (NotFoundException | NotModifiedException e) {
+            log.debug("Container with id {} is already stopped: {}", containerId, e.getMessage());
+        }
+        catch (Exception e) {
+            // In case the stopContainerCmd fails, we try to forcefully kill the container
+            try {
+                dockerClient.killContainerCmd(containerId).exec();
```

Comment:
Shouldn't we use `removeContainerCmd` here? This will only stop the container

Reply:
the build job containers are configured to auto-remove once stopped, so removeContainerCmd is not necessary